# M8: Excercise Solutions

## Excercise 1

The website describes that 14 out of 75 attributes of patients are being used for the classification task (determining whether the patient has a heart disease or not). 13 out of 14 of these values are features and the 14th is the target value. A description of each of them can be seen in the variables table.

## Excercise 2

In [41]:
# install the library using a shell command - you could also use the terminal
!pip install ucimlrepo

# import the library
from ucimlrepo import fetch_ucirepo

## Excercise 3

In [42]:
# fetch dataset 
heart_disease = fetch_ucirepo(id=45) 
# print the type of the instance
print(type(heart_disease))

<class 'ucimlrepo.dotdict.dotdict'>


## Excercise 4

In [43]:
# metadata 
metadata = heart_disease.metadata
print(metadata) 

{'uci_id': 45, 'name': 'Heart Disease', 'repository_url': 'https://archive.ics.uci.edu/dataset/45/heart+disease', 'data_url': 'https://archive.ics.uci.edu/static/public/45/data.csv', 'abstract': '4 databases: Cleveland, Hungary, Switzerland, and the VA Long Beach', 'area': 'Health and Medicine', 'tasks': ['Classification'], 'characteristics': ['Multivariate'], 'num_instances': 303, 'num_features': 13, 'feature_types': ['Categorical', 'Integer', 'Real'], 'demographics': ['Age', 'Sex'], 'target_col': ['num'], 'index_col': None, 'has_missing_values': 'yes', 'missing_values_symbol': 'NaN', 'year_of_dataset_creation': 1989, 'last_updated': 'Fri Nov 03 2023', 'dataset_doi': '10.24432/C52P4X', 'creators': ['Andras Janosi', 'William Steinbrunn', 'Matthias Pfisterer', 'Robert Detrano'], 'intro_paper': {'ID': 231, 'type': 'NATIVE', 'title': 'International application of a new probability algorithm for the diagnosis of coronary artery disease.', 'authors': 'R. Detrano, A. Jánosi, W. Steinbrunn, M

Answers to the questions should all be shown in the print statement:

- How many data points / patients do we have and how many features? 
    'num_instances': 303 --> 303 data points
    'num_features': 13 --> 13 features
- What are the names of the demographic features? 
    'demographics': ['Age', 'Sex'] --> age and sex are the demographic features
- What is the target value called?
    'target_col': ['num'] --> num is the name of the target
- Are there any missing values in the dataset and if yes how can they be identified?
    'has_missing_values': 'yes' --> there are missing values
    'missing_values_symbol': 'NaN' --> they can be identified by the NaN value

## Excercise 5

In [44]:
# variable information 
variables_information = heart_disease.variables
print(variables_information) 

        name     role         type demographic  \
0        age  Feature      Integer         Age   
1        sex  Feature  Categorical         Sex   
2         cp  Feature  Categorical        None   
3   trestbps  Feature      Integer        None   
4       chol  Feature      Integer        None   
5        fbs  Feature  Categorical        None   
6    restecg  Feature  Categorical        None   
7    thalach  Feature      Integer        None   
8      exang  Feature  Categorical        None   
9    oldpeak  Feature      Integer        None   
10     slope  Feature  Categorical        None   
11        ca  Feature      Integer        None   
12      thal  Feature  Categorical        None   
13       num   Target      Integer        None   

                                          description  units missing_values  
0                                                None  years             no  
1                                                None   None             no  
2              

Answers for the questions:
- which variables have missing values? --> 'ca' and 'thal'
- what is the unit of 'resting blood pressure'? --> mm Hg
- how many categorial variables are there? --> 7

## Excercise 6

In [45]:
# data (as pandas dataframes) 
X = heart_disease.data.features 
y = heart_disease.data.targets 

#import pandas
import pandas as pd
#combine the features and target into a single dataframe
df = pd.concat([X, y], axis=1)
#display the first 5 rows of the dataframe
print(df.head())

   age  sex  cp  trestbps  chol  fbs  restecg  thalach  exang  oldpeak  slope  \
0   63    1   1       145   233    1        2      150      0      2.3      3   
1   67    1   4       160   286    0        2      108      1      1.5      2   
2   67    1   4       120   229    0        2      129      1      2.6      2   
3   37    1   3       130   250    0        0      187      0      3.5      3   
4   41    0   2       130   204    0        2      172      0      1.4      1   

    ca  thal  num  
0  0.0   6.0    0  
1  3.0   3.0    2  
2  2.0   7.0    1  
3  0.0   3.0    0  
4  0.0   3.0    0  


## Excercise 7

In [46]:
#print 'ca' collumn from df
print('"Ca" column: ')
print(df['ca'])

#count number of NaN values in 'ca' collumn
print(f'Number of Nan values in "ca": {df['ca'].isna().sum()}')

# count number of unique values in 'ca' collumn
print(f'Number of unique values in "ca": {df['ca'].nunique()}')
# print the unique values in 'ca' collumn
print(f'Unique values in "ca": {df['ca'].unique()}')

# print data type of 'ca' from the variables_information
print(f'Data type of "ca": {variables_information[variables_information["name"] == "ca"]["type"].values[0]}')


#print 'thal' collumn from df
print('"Thal" column: ')
print(df['thal'])

#count number of NaN values in 'thal' collumn
print(f'Number of Nan values in "thal": {df['thal'].isna().sum()}')

# count number of unique values in 'thal' collumn
print(f'Number of unique values in "thal": {df['thal'].nunique()}')
# print the unique values in 'thal' collumn
print(f'Unique values in "thal": {df['thal'].unique()}')

# print data type of 'thal' from the variables_information
print(f'Data type of "thal": {variables_information[variables_information["name"] =="thal"]["type"].values[0]}')

"Ca" column: 
0      0.0
1      3.0
2      2.0
3      0.0
4      0.0
      ... 
298    0.0
299    2.0
300    1.0
301    1.0
302    NaN
Name: ca, Length: 303, dtype: float64
Number of Nan values in "ca": 4
Number of unique values in "ca": 4
Unique values in "ca": [ 0.  3.  2.  1. nan]
Data type of "ca": Integer
"Thal" column: 
0      6.0
1      3.0
2      7.0
3      3.0
4      3.0
      ... 
298    7.0
299    7.0
300    7.0
301    3.0
302    3.0
Name: thal, Length: 303, dtype: float64
Number of Nan values in "thal": 2
Number of unique values in "thal": 3
Unique values in "thal": [ 6.  3.  7. nan]
Data type of "thal": Categorical


## Excersice 8

In [47]:
# Remove rows with NaN values in 'ca' and 'thal' columns
print('Removing rows with NaN values in "ca" and "thal" columns.')
df = df.dropna(subset=['ca', 'thal'])

# Reset index after dropping rows
df.reset_index(drop=True, inplace=True)

# Print number NaN values for each column
print('Number of NaN values for each column:')
print(df.isna().sum())

# Print shape of the dataframe
print('Shape of the dataframe:')
print(df.shape)

Removing rows with NaN values in "ca" and "thal" columns.
Number of NaN values for each column:
age         0
sex         0
cp          0
trestbps    0
chol        0
fbs         0
restecg     0
thalach     0
exang       0
oldpeak     0
slope       0
ca          0
thal        0
num         0
dtype: int64
Shape of the dataframe:
(297, 14)


## Exercise 9

In [49]:
# Check if any of the rows are duplicates
print('Checking for duplicate rows:')
duplicates = df.duplicated()
print(f'Number of duplicates: {duplicates.sum()}')
print(f'Duplicate rows: {duplicates}')

# Check if any of the columns are duplicates
print('Checking for duplicate columns:')
duplicates = df.T.duplicated()
print(f'Duplicate columns: {duplicates}')
print(f'Number of duplicate columns: {duplicates.sum()}')

Checking for duplicate rows:
Number of duplicates: 0
Duplicate rows: 0      False
1      False
2      False
3      False
4      False
       ...  
292    False
293    False
294    False
295    False
296    False
Length: 297, dtype: bool
Checking for duplicate columns:
Duplicate columns: age         False
sex         False
cp          False
trestbps    False
chol        False
fbs         False
restecg     False
thalach     False
exang       False
oldpeak     False
slope       False
ca          False
thal        False
num         False
dtype: bool
Number of duplicate columns: 0


Excercise 10

In [52]:
# Get type of 'num' column
print(f'Data type of "num": {variables_information[variables_information["name"] == "num"]["type"].values[0]}') 
# Get unique values of 'num' column
print(f'Unique values of "num": {df["num"].unique()}')
# Get number of unique values of 'num' column
print(f'Number of unique values of "num": {df["num"].nunique()}')

# Add a new column 'heart_diseases_binary'. For each row, the entry is 0 if 'num' is 0 and 1 otherelse.
df['heart_diseases_binary'] = df['num'].apply(lambda x: 0 if x == 0 else 1)

# Print the first 5 rows of the df
print(df.head())

Data type of "num": Integer
Unique values of "num": [0 2 1 3 4]
Number of unique values of "num": 5
   age  sex  cp  trestbps  chol  fbs  restecg  thalach  exang  oldpeak  slope  \
0   63    1   1       145   233    1        2      150      0      2.3      3   
1   67    1   4       160   286    0        2      108      1      1.5      2   
2   67    1   4       120   229    0        2      129      1      2.6      2   
3   37    1   3       130   250    0        0      187      0      3.5      3   
4   41    0   2       130   204    0        2      172      0      1.4      1   

    ca  thal  num  heart_diseases_binary  
0  0.0   6.0    0                      0  
1  3.0   3.0    2                      1  
2  2.0   7.0    1                      1  
3  0.0   3.0    0                      0  
4  0.0   3.0    0                      0  
